pkg/controller/service/service_controller.go

```diff
@@ -267,12 +267,12 @@ func (s *ServiceController) processServiceCreateOrUpdate(service *v1.Service, ke
 	// Always cache the service, we need the info for service deletion in case
 	// when load balancer cleanup is not handled via finalizer.
 	cachedService.state = service
-	op, err := s.syncLoadBalancerIfNeeded(service, key)
+	op, err, loadBalancerDeleted := s.syncLoadBalancerIfNeeded(service, key)
 	if err != nil {
 		s.eventRecorder.Eventf(service, v1.EventTypeWarning, "SyncLoadBalancerFailed", "Error syncing load balancer: %v", err)
 		return err
 	}
-	if op == deleteLoadBalancer {
+	if op == deleteLoadBalancer && loadBalancerDeleted {
```

Comment:
What's the reason why we can't just check if `err == nil`?

Reply:
The reason is that error may be returned corresponding to removeFinalizer() call.
However, in that case, the load balancer has been removed. Hence the cache should be updated accordingly.

Reply:
If an error is returned by `removeFinalizer()`, it won't even reach this line?